File: Modern/src/main/scala/prv/zielony/transaction/analyst/modern/dao/AccountsDao.scala
package prv.zielony.transaction.analyst.modern.dao

import com.datastax.driver.core.Row
import com.websudos.phantom.dsl._
import prv.zielony.transaction.analyst.modern.model.UserAccount

import scala.concurrent.Future

/**
 * Created by zielony on 06.03.16.
 */
class AccountsDao extends CassandraTable[AbstractAccountsDao, UserAccount] {

  object id extends LongColumn(this) with PartitionKey[Long]

  object balance extends DoubleColumn(this)

  object userId extends LongColumn(this) with Index[Long]

  object userCountry extends OptionalStringColumn(this) with Index[Option[String]]

  override def fromRow(row: Row): UserAccount = {
    UserAccount(id(row), balance(row), userId(row), userCountry(row))
  }
}

abstract class AbstractAccountsDao extends AccountsDao with RootConnector {

  def findBalanceByCountry(country:String):Future[Seq[Double]] = {
    select(_.balance) where(_.userCountry eqs country) fetch
  }

  def countAll():Future[Option[Long]] = {
    select.count().one()
  }

  def persistUserAccount(account:UserAccount):Future[ResultSet] = {
    insert() 
      .value(_.id, account.id) 
      .value(_.balance, account.balance) 
      .value(_.userId, account.userId)
      .value(_.userCountry, account.userCountry)
      .future
  }
}

File: Classical/src/main/java/prv/zielony/transaction/analyst/classical/dao/AccountsDao.java
package prv.zielony.transaction.analyst.classical.dao;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;
import prv.zielony.transaction.analyst.classical.model.Account;

import java.time.LocalDateTime;
import java.util.List;

/**
 * Created by zielony on 26.02.16.
 */
@Repository
public interface AccountsDao extends JpaRepository<Account, Long> {

    @Query("SELECT a FROM Account a JOIN FETCH a.owner o WHERE D o.personalData IS NOT NULL " +
            "AND o.personalData.country IS NOT NULL AND o.personalData.country = :country AND o.birthDate IS NOT NULL " +
            "AND o.birthDate >= :startDate AND o.birthDate <= :endDate")
    public List<Account> findAccountsByCountryAndUserBirthDate(@Param("country") String country,
                                                               @Param("startDate") LocalDateTime startDate,
                                                               @Param("endDate") LocalDateTime endDate);

    @Query("SELECT COUNT(a) FROM Account a JOIN FETCH a.owner o WHERE o.personalData IS NOT NULL AND" +
            " o.personalData.country IS NOT NULL AND o.personalData.country = :country")
    public int countAccountsByCountry(@Param("country") String country);
}
